### PR TITLE
Recipe Timeline

### PR DESCRIPTION
**Pull request by mythmon**
I added a new page, and in doing so made a base component for new pages. That new page is a timeline of events for a recipe. I've been wanting to make this kind of thing to automate the analysis I did in mozilla/normandy#2040. The devtools seemed like a good place to do it.

image

### CODE REVIEW

File: extension/content/components/pages/AddonStudiesPage.js [hunk 1]
```diff
@@ -2,10 +2,11 @@ import React from "react";
 import { Loader } from "rsuite";
 
 import AddonStudy from "devtools/components/studies/AddonStudy";
+import BasePage from "./BasePage";
```
Comment (rehandalal) [inline extension/content/components/pages/AddonStudiesPage.js]: nit: use absolute import path

File: extension/experiments/normandy/api.js [hunk 2]
```diff
@@ -87,7 +87,12 @@ var normandy = class extends ExtensionAPI {
           },
 
           async checkRecipeFilter(recipe) {
-            return RecipeRunner.checkFilter(recipe);
+            if (RecipeRunner.checkFilter) {
+              return RecipeRunner.checkFilter(recipe);
+            } else if (RecipeRunner.shouldRunRecipe) {
+              return RecipeRunner.shouldRunRecipe(recipe);
+            }
+            throw new Error("Incompatible version of Normandy");
```
Comment (rehandalal) [inline extension/experiments/normandy/api.js:95]: 👏🏽 one thing i'd love to look at as we grow this addon is regression testing against nightly to catch stuff like this early.

File: .eslintrc.js [hunk 3]
```diff
@@ -9,5 +9,6 @@ module.exports = {
   ],
   rules: {
     "mozilla/no-define-cc-etc": "off", // seems broken outside of m-c
+    "generator-star-spacing": "off",
```
Comment (rehandalal) [inline .eslintrc.js]: can we pick something for a consistent spacing of the star rather than turning this off?

Comment (rehandalal) [reply .eslintrc.js]: i think im slightly partial to `both`

File: extension/content/components/pages/AddonStudiesPage.js [hunk 4]
```diff
@@ -2,10 +2,11 @@ import React from "react";
...
 
 const normandy = browser.experiments.normandy;
 
-export default class PrefStudiesPage extends React.PureComponent {
+export default class PrefStudiesPage extends BasePage {
```
Comment (rehandalal) [inline extension/content/components/pages/AddonStudiesPage.js]: the official take on inheritance: https://reactjs.org/docs/composition-vs-inheritance.html

File: extension/content/components/pages/RecipesPage.js [hunk 5]
```diff
@@ -1,170 +1,66 @@
...
-  Pagination,
-  SelectPicker,
-} from "rsuite";
+import { Button, Icon, Loader, Modal, Nav, Pagination } from "rsuite";
+import { Map } from "immutable";
 
 import { ENVIRONMENTS } from "devtools/config";
 import RecipeListing from "devtools/components/recipes/RecipeListing";
-import api from "devtools/utils/api";
+import BaseApiPage from "./BaseApiPage";
```
Comment (rehandalal) [inline extension/content/components/pages/RecipesPage.js]: nit: no relative import paths

File: extension/content/components/recipes/RecipeTimeline.js [hunk 6]
```diff
@@ -0,0 +1,126 @@
+import React from "react";
+import PropTypes from "prop-types";
+import { Timeline, Icon } from "rsuite";
+import moment from "moment";
+
+import "./RecipeTimeline.less";
```
Comment (rehandalal) [inline extension/content/components/recipes/RecipeTimeline.js]: nit: use absolute

**Review by rehandalal — CHANGES_REQUESTED** [review]

File: .eslintrc.js (hunk 3 above)
Comment (mythmon) [reply .eslintrc.js]: Prettier (from `extends: [..., "plugin:prettier/recommended"]`) handles generator star spacing. This is just getting ESLint to stop fighting with it. This ESLint rule should have been disabled by the ESLint/Prettier integration, I'm not sure why it wasn't.

File: extension/content/components/pages/AddonStudiesPage.js (hunk 4 above)
Comment (mythmon) [reply extension/content/components/pages/AddonStudiesPage.js]: I'll switch it over.

**Review by mythmon — COMMENTED** [review]
Ack, sorry about the relative imports. VSCode has some very handy automatic import logic, but it's sort of invisible and it makes relative imports. I'll clean those up.

**Comment by mythmon** [comment]
This is sort of ready to go. If you have time to review it, go for it. I think I might split this into two PRs though, one for the basepage stuff and one for the timeline page. I think that will make things simpler overall.

**Comment by mythmon** [comment]
I'm actually going to abandon this. I don't think it's worth fixing the merge conflicts and cleaning this up. I think the idea of a common base page is a good idea though. That's probably something for the editing refactor though. I think the idea of a base page is something we should use in the future.

**Pull request closed without merging**